flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -462,12 +486,20 @@ public final void invoke() throws Exception {
 					operatorChain.releaseOutputs();
 				}
 			}
+
+			// close mailbox
+			FutureUtils.cancelRunnableFutures(taskMailboxExecutor.shutdownNow());
 		}
 	}
 
 	@Override
 	public final void cancel() throws Exception {
-		mailbox.clearAndPut(POISON_LETTER);
+		try {
+			List<Runnable> droppedRunnables = taskMailboxExecutor.getMailbox().clearAndPut(mailboxPoisonLetter);
+			FutureUtils.cancelRunnableFutures(droppedRunnables);
```

Comment:
At the moment the pending letters are not under real execution although they may be a future instance, why needs to do the cancel operation on them ?

Reply:
Currently this is done to signal interested clients that their letter will never be executed. I think this can be useful for shutdown in some cases.